service/raft/raft_group0.cc

```diff
@@ -858,31 +858,39 @@ future<> raft_group0::become_nonvoter(abort_source& as, std::optional<raft_timeo
     auto my_id = load_my_id();
     group0_log.info("becoming a non-voter (my id = {})...", my_id);
 
-    co_await modify_raft_voter_status({my_id}, can_vote::no, as, timeout);
+    co_await modify_raft_voter_config({}, {my_id}, as, timeout);
     group0_log.info("became a non-voter.", my_id);
 }
 
-future<> raft_group0::set_voter_status(raft::server_id node, can_vote can_vote, abort_source& as, std::optional<raft_timeout> timeout) {
-    co_return co_await set_voters_status({node}, can_vote, as, timeout);
+future<> raft_group0::make_nonvoter(raft::server_id node, abort_source& as, std::optional<raft_timeout> timeout) {
+    co_return co_await modify_voters({}, {node}, as, timeout);
 }
 
-future<> raft_group0::set_voters_status(
-        const std::unordered_set<raft::server_id>& nodes, can_vote can_vote, abort_source& as, std::optional<raft_timeout> timeout) {
-    if (!(co_await raft_upgrade_complete())) {
-        on_internal_error(group0_log, "called set_voter_status before Raft upgrade finished");
-    }
-
-    if (nodes.empty()) {
+future<> raft_group0::modify_voters(const std::unordered_set<raft::server_id>& nodes_add, const std::unordered_set<raft::server_id>& nodes_del,
+        abort_source& as, std::optional<raft_timeout> timeout) {
+    if (nodes_add.empty() && nodes_del.empty()) {
         co_return;
     }
 
-    const std::string_view status_str = can_vote ? "voters" : "non-voters";
+    if (!(co_await raft_upgrade_complete())) {
+        on_internal_error(group0_log, "called modify_voters before Raft upgrade finished");
+    }
 
-    group0_log.info("making servers {} {}...", nodes, status_str);
+    if (!nodes_add.empty()) {
+        group0_log.info("making servers {} voters ...", nodes_add);
+    }
+    if (!nodes_del.empty()) {
+        group0_log.info("making servers {} non-voters ...", nodes_del);
+    }
 
-    co_await modify_raft_voter_status(nodes, can_vote, as, timeout);
+    co_await modify_raft_voter_config(nodes_add, nodes_del, as, timeout);
 
-    group0_log.info("servers {} are now {}.", nodes, status_str);
+    if (!nodes_add.empty()) {
+        group0_log.info("servers {} are now voters.", nodes_add);
+    }
+    if (!nodes_del.empty()) {
+        group0_log.info("servers {} are now non-voters.", nodes_del);
+    }
```

Comment:
Logging a few INFO logs per topology operation/shutdown/restart should be fine, but perhaps the voter registry would be a better place, I'm not sure. Let's at least ensure we don't log the same information in two places.

Reply:
OK